### ISSUE TITLE
Unexpected void type in the formula

### ISSUE THREAD

**Issue by Robulane** [microsoft/Power-Fx#2221]
This test fails

        [Fact]
        public void CheckParseSuccess()
        {
            var config = new PowerFxConfig();
            var engine = new Engine(config);
            var parse = engine.Parse("If(Len(x) = 0, y, x)");

            var r = RecordType.Empty()
                .Add(new NamedFormulaType("x", FormulaType.String))
                .Add(new NamedFormulaType("y", FormulaType.UntypedObject));

            var check = engine.Check(parse, r);
            Assert.True(check.IsSuccess);

            Assert.IsType<UntypedObjectType>(check.ReturnType);
        }

**Comment by Robulane** [comment]
I disagree that this is expected. The Void should never be returned for the expression that expects the value. You should either return the error or the type of the first argument. If the first argument is Untyped, I think you can return Untyped (because it's untyped). Expressions like this should return an error though: If(true, 2, "")

**Comment by Robulane** [comment]
We should not wait the Runtime to throw an error, the check function should return the error.

**Comment by MikeStall** [comment]
For imperative, void is useful. 
For expression-only, void should be an error. 

Fortunately, we do know whether an expression allows side-effects. This is determined via the parser options, and it impacts things like whether we have have chaining operating  or call behavior functions. 

Proposal: if the expression does not allow side-effects, then any usage of void becomes an error from power fx.
